Support dynamically creating tailored continuation layouts

jakobbotsch opened the pull request:
Add ability for the VM to dynamically create continuation layout types and for the JIT to request such types to be created.

jakobbotsch committed:
Initial

jakobbotsch committed:
Fix build, make tests run

jakobbotsch committed:
Tailored continuation layouts

Add ability for the VM to dynamically create continuation layout types
and for the JIT to request such types to be created.

jakobbotsch committed:
Update type

jakobbotsch committed:
Fix bug

jakobbotsch committed:
Fix some bugs

jakobbotsch committed:
Run thunk generator

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10254,6 +10252,31 @@ void CEEInfo::getAsyncInfo(CORINFO_ASYNC_INFO* pAsyncInfoOut)
+
+    AllocMemTracker amTracker;
+    // TODO: table to share/deduplicate these
+    result = (CORINFO_CLASS_HANDLE)m_pMethodBeingCompiled->GetModule()->CreateContinuationMethodTable((unsigned)dataSize, objRefs, dataOffsets, &amTracker);
```
I think these should be created on the loader allocator.

jakobbotsch replied:
That makes a lot of sense. Will change this.

davidwrighton replied:
The parts of the system that will likely get in trouble are if we (via reflection) get a System.Type of one of these, and the various debugger and profiler apis which examine the heap and various values. Please get @tommcdon or @noahfalk to take a look at this before we signoff.

jakobbotsch replied:
Gotcha. At least that is some evidence that things may not be _badly_ broken with these new types. I suppose they still differentiate themselves in that they have no metadata token available, so that may cause some problems of the "badly broken" kind.

I have opened #120800 about follow-up work here and quoted some of the comments from this comment thread.

jkotas commented on src/coreclr/vm/methodtable.h:
```diff
@@ -2788,6 +2788,19 @@ class MethodTable
+    inline BOOL IsContinuation()
+    {
+        LIMITED_METHOD_DAC_CONTRACT;
+        return GetFlag(enum_flag_IsContinuationType);
```
Can this check whether the parent type is a `CLASS__CONTINUATION` instead?

jakobbotsch replied:
Yeah, that seems better, no need to add this flag

jakobbotsch commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs:
```diff
@@ -49,81 +49,80 @@ public void Pop()
-            Debug.Assert(continuationContext != null);
-            return continuationContext;
+            // Only the special continuation sub types are expected.
+            Debug.Assert(GetType() != typeof(Continuation));
```
We currently create an instance of `Continuation` inside `FinalizeTaskReturningThunk`, though I think with change and some more refactoring we can probably get rid of that.

jakobbotsch replied:
Hmm, we also create one in `AwaitAwaiter` and `UnsafeAwaitAwaiter`. That one is harder to remove.

jakobbotsch committed:
Create a manager in the LoaderAllocator to hold map and create types

jakobbotsch committed:
Use SPC for continuation's module

jakobbotsch committed:
Remove unnecessary continuation allocation

Now that we store directly into the ThunkTask's result we do not need to
allocate a tail continuation at all.

jakobbotsch committed:
Fix some bugs

jakobbotsch commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10254,6 +10252,32 @@ void CEEInfo::getAsyncInfo(CORINFO_ASYNC_INFO* pAsyncInfoOut)
+    LoaderAllocator* allocator = m_pMethodBeingCompiled->GetLoaderAllocator();
+    AsyncContinuationsManager* asyncConts = allocator->GetAsyncContinuationsManager();
+    AllocMemTracker amTracker;
+    result = (CORINFO_CLASS_HANDLE)asyncConts->LookupOrCreateContinuationMethodTable((unsigned)dataSize, objRefs, dataOffsets, m_pMethodBeingCompiled, &amTracker);
```
Strictly speaking we can delay this call until the continuation needs to be created. But in any case we need to persist some description of the continuation somewhere. However, we would be able to store that description compressed until it is needed.
I don't have good intuition about the expense of creating these or the number of them that we are going to be creating (likely a lot), so it is hard to know if this will be good enough or if it will show up on the radar.

jakobbotsch committed:
Fix type GC info, add debug name, remove comments

jakobbotsch committed:
Refactor building of obj refs bitmap into helper class

jakobbotsch committed:
Properly allocate space for KeepAlive object

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into tailored-continuation-layouts

jakobbotsch committed:
Fix bitmap building for implicit byrefs

jakobbotsch commented on src/coreclr/vm/methodtable.inl:
```diff
@@ -299,6 +299,13 @@ inline BOOL MethodTable::IsValueType()
+{
+    LIMITED_METHOD_DAC_CONTRACT;
+
+    return m_pParentMethodTable == CoreLibBinder::GetClassIfExist(CLASS__CONTINUATION);
```
Probably cannot use `GetClassIfExist` here, it looks quite expensive to repeatedly call if the class hasn't been loaded. I suppose I can stash the type in a static when we create one of the continuation sub types.

jakobbotsch committed:
Run jit-format, remove copilot comment

jakobbotsch committed:
Move Continuation MT into a static

jakobbotsch committed:
Remove added newline

jakobbotsch committed:
Undo another change

jakobbotsch committed:
Undo name change

jakobbotsch committed:
Add a cache for created layout types

jakobbotsch committed:
Fix a bug

jakobbotsch committed:
Fix another bug

jakobbotsch committed:
Fix debug counting

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into tailored-continuation-layouts

jakobbotsch committed:
Fix MethodTable::ValidateWithPossibleAV for new types

jakobbotsch committed:
Add assert in `GetModule()`, make `GetType()` return `typeof(Continuation)`

jakobbotsch committed:
Minor cleanups

jakobbotsch committed:
Throw `NotSupportedException` for Object.GetType() called on continuations

jakobbotsch committed:
Update src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into tailored-continuation-layouts

jakobbotsch committed:
Handle large alignment requirements

jakobbotsch committed:
Fix after merge

jakobbotsch committed:
Allocate singleton subclassed continuation with a corresponding `EEClass`

jakobbotsch committed:
Enable runtime async by default for testing

noahfalk commented on src/coreclr/vm/asynccontinuations.h:
```diff
@@ -0,0 +1,61 @@
+
+public:
+    AsyncContinuationsManager(LoaderAllocator* allocator);
+    MethodTable* LookupOrCreateContinuationMethodTable(unsigned dataSize, const bool* objRefs, const CORINFO_CONTINUATION_DATA_OFFSETS& dataOffsets, MethodDesc* asyncMethod);
```
> What name should we be emitting in these events? Currently the name will be (dynamicClass). For debug purposes I am creating a name that looks like Continuation_[inducing method]_[size]_[number of GC refs] to put in debug_m_szClassName. But I am not sure we would want to store a name like that. Maybe (dynamicContinuation) would be enough. Thoughts?

`Continuation_[inducing method]_[size]_[number of GC refs]` sounds fine to me. I don't think the exact name we pick is critical but it would be good for us to be consistent in different places that we do give it a name and make the name distinct enough that it can distinguish the different types. You mentioned storing it but would we need to? I was thinking we have the underlying pieces of info available and we can format the name string on-demand if the ETW events are on.

> This one seems to skip array types. Is it reasonable to do the same for continuation types, or should we fire them?

While we could get away not firing them, assuming its not much work I think we are better off adding them now. If we don't add them now and then find scenario where they would be more useful in the future we either have to do another breaking change or add yet more opt-in switches. I'd rather get these things exposed in one go.

> I think [the loader allocator heap] is good enough?

Yep, that should cover it! You were right that I was just referring to the dictionary as the thing which transitively would capture the Continuation types. Capturing them via the heap works fine.

jkotas replied:
> `Continuation_[size]_[number of GC refs]`

"number of GC refs" is ambiguous. It won't capture different ordering of GC refs. Should it rather capture the GC info into non-ambiguous way so that the name is unique when there are no collectible assemblies in the picture?

jakobbotsch replied:
Sure, I can encode all the GC series to make it unambiguous.

jakobbotsch replied:
If we call `ClassLoader::NotifyLoad` for these, then I assume we must also call `ClassLoader::NotifyUnload`. Is it fine to just call that from `LoaderAllocator::Terminate`? For other classes it seems we call it from `Module::FreeClassTables`, but since these are associated with the `LoaderAllocator` I am not 100% sure when to do the notification.

jakobbotsch replied:
I pushed a commit for this, but I still need to do some more testing.

jkotas replied:
> Mostly I am suggesting we add them just to get the breaking change potential out of the way

Should we add them for arrays too then? It does not make sense for these notifications to be explicitly skipped for arrays and be emitted for continuations.

noahfalk replied:
Yeah, turning on the arrays at the same time makes sense to me.

jakobbotsch committed:
Fix bug

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Give continuations names, fire profiler/ETW events

jakobbotsch committed:
Get Continuation class name from EE

jakobbotsch committed:
Fire duplicate TypeLoadEnd events instead of no matching pair

jakobbotsch committed:
Fix non Windows build

jkotas commented on src/coreclr/vm/asynccontinuations.cpp:
```diff
@@ -243,23 +258,41 @@ MethodTable* AsyncContinuationsManager::LookupOrCreateContinuationMethodTable(un
+        }
+    }
+
+    if (pReturnedMT == pNewMT)
```
Right. Also, the regular types are loaded in multiple stages (`ClassLoadLevel` enum) so you can see multiple events for a single type, each on different thread.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10254,6 +10252,35 @@ void CEEInfo::getAsyncInfo(CORINFO_ASYNC_INFO* pAsyncInfoOut)
+
+    LoaderAllocator* allocator = m_pMethodBeingCompiled->GetLoaderAllocator();
+    AsyncContinuationsManager* asyncConts = allocator->GetAsyncContinuationsManager();
+    result = (CORINFO_CLASS_HANDLE)asyncConts->LookupOrCreateContinuationMethodTable((unsigned)dataSize, objRefs, dataOffsets, m_pMethodBeingCompiled);
```
Do you see it as a temporary solution or permanent design - is our strategy to always copy everything from the stack to the continuation and back?

For example, if one has a code like this:
```
async Task M()
{
      LargeStructWithAlignedField parser = new();
      while (!parser.Done)
      {
            await parser.Next();
            ...
      }
}
```
Do we plan to copy the whole `parser` back and forth?

jakobbotsch replied:
Yes, today we would copy `parser` back and forth.

I don't think anything in the current design and spec precludes us from a different strategy that optimizes for suspend/resume performance. That would likely involve allocating a "mega-continuation" ahead of time and the JIT rewriting things early on during compilation so that all local accesses are loads/stores from this class. Basically the asyncv1 strategy. In that case we would want to support the double alignment of these mega-continuations. But the JIT is not good at optimizing this kind of code, even if strictly speaking it can know that it has private access to the `Continuation` object. Building our optimizations out to support this scenario would be a large undertaking in and of itself.

It could also be a hybrid strategy that does so piecewise -- like in your case, if we determine `parser` is sufficiently large and that we are going to be suspending/resuming often with it live, then we could decide to box it ahead of time.

BTW, for these hybrid strategies we would need to think more about compatibility with https://github.com/dotnet/csharplang/blob/main/proposals/csharp-13.0/ref-unsafe-in-iterators-async.md. Even today I am not sure this C# feature is compatible with the wording of the current specification, even if it is compatible with how the current implementation works.

jkotas commented on src/coreclr/vm/typehandle.cpp:
```diff
@@ -318,6 +325,12 @@ bool TypeHandle::IsManagedClassObjectPinned() const
+    {
+        *pDest = OBJECTREFToObject(GetParent().GetManagedClassObject());
+        return;
+    }
```
It should not be possible in (safe) user code to get a reference to the continuation. It is likely going to be possible under debugger in eval window.

jakobbotsch replied:
I assume safe user code excludes reflection access of private fields?

When a chain of runtime async methods suspend the head `Continuation` instance is stored in `RuntimeAsyncTask.m_stateObject`, and the `RuntimeAsyncTask` instance is the `Target` of the continuation delegates passed to `INotifyCompletion.OnComplete` user code. So the instance can be retrieve through that path.

jakobbotsch committed:
Fix OFFSETOF__CORINFO_Continuation__data for 32 bit

jakobbotsch committed:
Remove CORINFO_CONTINUATION_DATA_OFFSETS; implement SuperPMI support

jakobbotsch committed:
Sort all object refs to be first

jakobbotsch committed:
Pass keep alive offset as an argument to complex allocation helpers

jakobbotsch commented:
> Test failures look related: System.InvalidProgramException: Invalid IL code in AwaitNotAsync:AsyncEntryPoint

I think this one was just Mono complaining about not supporting runtime async. But the 32-bit coreclr failures were related. Pushed a fix for that one, and I need to remember to revert the enabling of the testing.

jakobbotsch commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs:
```diff
@@ -49,81 +49,82 @@ public void Pop()
+            MethodTable* mt = RuntimeHelpers.GetMethodTable(this);
+
+            // Only the special continuation sub types have continuation offsets.
+            Debug.Assert(mt->IsContinuation);
```
Pushed a commit to remove `CORINFO_CONTINUATION_DATA_OFFSETS`

jakobbotsch committed:
Remove managed MethodTable.IsContinuation

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs:
```diff
@@ -49,81 +50,65 @@ public void Pop()
-            object? continuationContext = GCData[index];
-            Debug.Assert(continuationContext != null);
-            return continuationContext;
+            Debug.Assert((Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_HAS_CONTINUATION_CONTEXT) != 0);
```
(preexisting problem)

`CORINFO_...` types and constants are meant to be exchange types on JIT/EE interface and they are not meant to leak to runtime. It is odd to see them in CoreLib code.

(When these types leak to runtime, it often causes confusion. #116675 fixed one of such leaks that caused confusion for interpreter v-team.)

jakobbotsch replied:
Is it strictly the name of the enum/enum members on the C# side you are thinking about, or do you also think the values of the enum members should be communicated to the JIT by the VM in a different way?

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs:
```diff
@@ -49,81 +50,65 @@ public void Pop()
+        private const int POINTER_SIZE = 4;
+#endif
+
+        private const int CONTINUATION_OFFSET_TO_DATA = POINTER_SIZE /* Next */ + POINTER_SIZE /* Resume */ + 8 /* Flags + State */;
```
```suggestion
        private const int DataOffset = PointerSize /* Next */ + PointerSize /* Resume */ + 8 /* Flags + State */;
```
Follow C# naming convention?

jakobbotsch replied:
Fixed this

jkotas commented on src/coreclr/inc/corinfo.h:
```diff
@@ -1700,27 +1700,23 @@ struct CORINFO_EE_INFO
     // if an exception is being propagated, needs to be resumed. The exception
     // should be placed at index 0 or 1 depending on whether the continuation
     // also expects a result.
     CORINFO_CONTINUATION_NEEDS_EXCEPTION = 2,
```
Naming nit: All other enum members are `HAS_...,` this one is `NEEDS_... `. Should this one be called HAS_EXCEPTION?

noahfalk commented on src/coreclr/inc/clrconfigvalues.h:
```diff
@@ -720,7 +720,7 @@ RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableRiscV64Zbb,             W("EnableRiscV64
 
 // Runtime-async
-RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_RuntimeAsync, W("RuntimeAsync"), 0, "Enables runtime async method support")
+RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_RuntimeAsync, W("RuntimeAsync"), 1, "Enables runtime async method support")
```
I didn't think we were quite ready to have this on by default in previews. For example Thays has some bug fixes to make breakpoints work and that seems like functionality even someone using a preview would expect to have functional.

jkotas replied:
The main reason that prevents us from turning the runtime config switch on is that it breaks other runtime features (e.g. COM interop). Tracked by https://github.com/dotnet/runtime/issues/115094 that needs to be fixed before the runtime config switch is turned on.

noahfalk reviewed (commented):
This looks good to me modulo making RuntimeAsync on by default. If there is a need to get it on by default regardless of the diagnostics I'd suggest we do it in a separate PR.

jakobbotsch committed:
Address feedback

jakobbotsch commented on src/coreclr/inc/corinfo.h:
```diff
@@ -1700,27 +1700,23 @@ struct CORINFO_EE_INFO
-    // placed in the GCData array at index 0. Not set if the callee is void.
-    CORINFO_CONTINUATION_RESULT_IN_GCDATA = 1,
+    CORINFO_CONTINUATION_HAS_OSR_ILOFFSET = 1,
     // If this bit is set the continuation resumes inside a try block and thus
```
Updated comments in both enums

AndyAyersMS reviewed (commented):
JIT changes look good.

jkotas commented on src/coreclr/inc/corinfo.h:
```diff
@@ -1735,10 +1740,6 @@ struct CORINFO_ASYNC_INFO
-    CORINFO_FIELD_HANDLE continuationGCDataFldHnd;
     // Whether or not the continuation needs to be allocated through the
     // helper that also takes a method handle
     bool continuationsNeedMethodHandle;
```
`bool continuationsNeedMethodHandle` is no longer needed and it can be deleted. The continuation type is attached to the collectible loader allocator now. It will prevent the loader allocator from unloading and we do not need to keep it alive explicitly.

(The shared generic handling done on the JIT side right before `continuationsNeedMethodHandle` is checked is still needed.)

jkotas reviewed (approved):
LGTM!

noahfalk reviewed (approved):
Ditto!

VSadov commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -14495,9 +14541,11 @@ static Signature BuildResumptionStubSignature(LoaderAllocator* alloc, AllocMemTr
     sigBuilder.AppendElementType(ELEMENT_TYPE_OBJECT); // return type
     sigBuilder.AppendElementType(ELEMENT_TYPE_OBJECT); // continuation
+    sigBuilder.AppendElementType(ELEMENT_TYPE_BYREF); // result location
+    sigBuilder.AppendElementType(ELEMENT_TYPE_U1);
```
What is in the last argument?

jakobbotsch replied:
This is the encoding for `ref byte` as `[byref, U1]`

VSadov commented on src/coreclr/vm/corelib.h:
```diff
@@ -720,10 +720,10 @@ DEFINE_METHOD(ASYNC_HELPERS,      ALLOC_CONTINUATION_METHOD, AllocContinuationMe
-DEFINE_METHOD(ASYNC_HELPERS,      FINALIZE_TASK_RETURNING_THUNK_1, FinalizeTaskReturningThunk, GM_Continuation_RetTaskOfT)
-DEFINE_METHOD(ASYNC_HELPERS,      FINALIZE_VALUETASK_RETURNING_THUNK, FinalizeValueTaskReturningThunk, SM_Continuation_RetValueTask)
-DEFINE_METHOD(ASYNC_HELPERS,      FINALIZE_VALUETASK_RETURNING_THUNK_1, FinalizeValueTaskReturningThunk, GM_Continuation_RetValueTaskOfT)
+DEFINE_METHOD(ASYNC_HELPERS,      FINALIZE_TASK_RETURNING_THUNK, FinalizeTaskReturningThunk, SM_RetTask)
```
Is AllocContinuationResultBox above still used?

VSadov commented on src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs:
```diff
@@ -274,39 +253,23 @@ void ITaskCompletionAction.Invoke(Task completingTask)
-                    }
-
-                    return task.TrySetResult(result);
+                    return task.TrySetResult(task.m_result);
```
This will assign `m_result = m_result`. That seems ok, but could be optimized a tiny bit in the future, since the result is already there.

VSadov reviewed (approved):
LGTM. Thanks!!

jakobbotsch committed:
Remove AllocContinuationResultBox

jakobbotsch committed:
Remove CORINFO_ASYNC_INFO::continuationsNeedMethodHandle

jakobbotsch committed:
Update pseudocode

jakobbotsch commented:
Some numbers for the micro benchmark in https://github.com/dotnet/runtime/issues/120865#issuecomment-3418452717 that measures raw suspension/resumption performance, with 10 million iterations:

Before the change: Took 596.9 ms
After the change: Took 558.0 ms
Async1 for comparison: Took 332.1 ms
